hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/AbstractLeafQueue.java

```diff
@@ -1264,7 +1264,10 @@ public CSAssignment assignContainers(Resource clusterResource,
         userAssignable = canAssignToUser(clusterResource, application.getUser(),
             userLimit, application, candidates.getPartition(),
             currentResourceLimits);
-        if (!userAssignable && Resources.fitsIn(cul.reservation, appReserved)) {
+        if (!userAssignable
+            && Resources.fitsIn(cul.reservation, appReserved)
+            // Consider updating cul only if the application attempt is active.
+            && applicationAttemptMap.containsKey(application.getApplicationAttemptId())) {
```

Comment:
applicationAttemptMap.containsKey(application.getApplicationAttemptId()) can be in the begin of  Iterator

Reply:
what happens when the application finishes when its past requests are in middle of iterator code block ?
Isn't it much more safer to do here? I think we can have one check in the beginning as well to skip unwanted calls.

Any thoughts ?